File: packages/core/src/evaluator.ts
import { TileState, TileResult, GuessResult } from './types';

export function evaluateGuess(solution: string, guess: string): GuessResult {
  const solutionUpper = solution.toUpperCase();
  const guessUpper = guess.toUpperCase();

  if (guessUpper.length !== solutionUpper.length) {
    throw new Error(`Guess length ${guessUpper.length} does not match solution length ${solutionUpper.length}`);
  }

  const tiles: TileResult[] = [];
  const solutionChars = solutionUpper.split('');
  const guessChars = guessUpper.split('');
  const used: boolean[] = new Array(solutionChars.length).fill(false);

  for (let i = 0; i < guessChars.length; i++) {
    if (guessChars[i] === solutionChars[i]) {
      tiles.push({ letter: guessChars[i], state: TileState.CORRECT });
      used[i] = true;
    } else {
      tiles.push({ letter: guessChars[i], state: TileState.ABSENT });
    }
  }

  for (let i = 0; i < guessChars.length; i++) {
    if (tiles[i].state === TileState.CORRECT) continue;

    for (let j = 0; j < solutionChars.length; j++) {
      if (!used[j] && guessChars[i] === solutionChars[j]) {
        tiles[i].state = TileState.PRESENT;
        used[j] = true;
        break;
      }
    }
  }

  const isCorrect = tiles.every(t => t.state === TileState.CORRECT);
  return { tiles, isCorrect };
}

File: apps/ios/Sources/Core/Evaluator.swift
import Foundation

public func evaluateGuess(solution: String, guess: String) -> GuessResult {
    let solutionUpper = solution.uppercased()
    let guessUpper = guess.uppercased()

    guard solutionUpper.count == guessUpper.count else {
        fatalError("Guess length \(guessUpper.count) does not match solution length \(solutionUpper.count)")
    }

    let solutionChars = Array(solutionUpper)
    let guessChars = Array(guessUpper)
    var tiles = [TileResult]()
    var used = [Bool](repeating: false, count: solutionChars.count)

    // First pass: mark CORRECT
    for i in 0..<guessChars.count {
        if guessChars[i] == solutionChars[i] {
            tiles.append(TileResult(letter: String(guessChars[i]), state: .correct))
            used[i] = true
        } else {
            tiles.append(TileResult(letter: String(guessChars[i]), state: .absent))
        }
    }

    // Second pass: mark PRESENT
    for i in 0..<guessChars.count {
        if tiles[i].state == .correct { continue }
        for j in 0..<solutionChars.count {
            if !used[j] && guessChars[i] == solutionChars[j] {
                tiles[i].state = .present
                used[j] = true
                break
            }
        }
    }

    let isCorrect = tiles.allSatisfy { $0.state == .correct }
    return GuessResult(tiles: tiles, isCorrect: isCorrect)
}
